### ISSUE TITLE
"No VM found" - volume provision failed

### ISSUE THREAD

**Issue by beanallergy** [juju-solutions/charm-vsphere-integrator#1]
Application "vsphere-integrator" installed successfully from the charm store. 
I followed the instructions from here: https://github.com/juju-solutions/charm-vsphere-integrator/blob/master/README.md , just changed the name of the storageclass from mystorage -> vsphere. I tried to set up first with juju trust, then manually set the vSphere credentials, all result in the same following error.

**Problem:** PVC status is stuck on "Pending". Describe pvc testclaim:
`
Name:          testclaim
Namespace:     default
StorageClass:  vsphere
Status:        Pending
Volume:
Labels:        <none>
Annotations:   kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"v1","kind":"PersistentVolumeClaim","metadata":{"annotations":{},"name":"testclaim","namespace":"default"},"spec":{"accessModes":["ReadWr...
               volume.beta.kubernetes.io/storage-provisioner=kubernetes.io/vsphere-volume
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:
Access Modes:
Events:
  Type     Reason              Age               From                         Message
  ----     ------              ----              ----                         -------
  Warning  ProvisioningFailed  5s (x4 over 23s)  persistentvolume-controller  Failed to provision volume with StorageClass "vsphere": No VM found
`

Juju version:
`
2.4.3-xenial-amd64
`

Kubernetes version:
`
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.2", GitCommit:"bb9ffb1654d4a729bb4cec18ff088eacc153c239", GitTreeState:"clean", BuildDate:"2018-08-07T23:17:28Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"12+", GitVersion:"v1.12.0-beta.1", GitCommit:"553515b823634919e731774556fae554c95a6a5f", GitTreeState:"clean", BuildDate:"2018-09-05T03:30:08Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
`

**Comment by kwmonroe** [comment]
@beanallergy I've seen the "No VM found" message before, and it came about because the k8s vsphere provider could not detect a valid id on the k8s-worker nodes.

In 1.12, the vsphere provider determines whether or not a node is valid by looking at the `ProviderID`.  Check to see it looks something like this:
```bash
$ juju ssh kubernetes-worker/0 '/snap/bin/kubectl describe no | grep ProviderID'
ProviderID:                  vsphere://420BE254-DDAF-2499-6337-F58EDC92EF5B
```

Next, PVCs only work if the VM has `disk.enableUUID=true`.  You can verify this using [govc](https://github.com/vmware/govmomi/blob/master/govc/README.md):
```bash
$ ./govc_darwin_amd64 vm.info -e=true juju-dc32dd-1 | grep disk
    disk.enableUUID:                  true
```

Or using vCenter, follow **Step 1: Enable disk UUID on Node virtual machines** in [sunek's blog](https://sunekjaergaard.blogspot.com/2018/02/making-canonical-distribution-of.html).

Finally, if all that looks good, restart the k8s controller manager and verify it's loading the vsphere volume plugin:
```
$ juju ssh kubernetes-master/0 'sudo systemctl restart snap.kube-controller-manager.daemon'
$ sleep 10
$ juju ssh kubernetes-master/0 'journalctl -u snap.kube-controller-manager.daemon | grep -i vsphere'
Sep 13 14:14:23 juju-dc32dd-0 kube-controller-manager.daemon[15523]: I0913 14:14:23.039150   15523 flags.go:27] FLAG: --cloud-provider="vsphere"
Sep 13 14:14:39 juju-dc32dd-0 kube-controller-manager.daemon[15523]: W0913 14:14:39.882608   15523 vsphere.go:298] SecretName and/or SecretNamespace is not provided. VCP will use username and password from config file
Sep 13 14:14:39 juju-dc32dd-0 kube-controller-manager.daemon[15523]: I0913 14:14:39.890876   15523 plugins.go:496] Loaded volume plugin "kubernetes.io/vsphere-volume"
```

**Comment by beanallergy** [comment]
Thanks @kwmonroe  for the really helpful follow-up. I found out you did cover this prerequisite in your blog here https://blog.ubuntu.com/2018/08/23/vsphere , would be nice if this is also included in the README. 

After I followed your instructions with govc, I still have to set providerID manually to each node spec (seems to be an ongoing issue https://github.com/kubernetes/kubernetes/issues/65933#issuecomment-408358655 ) Now I have a different error in testing a new pvc

` 'ProvisioningFailed' Failed to provision volume with StorageClass "vsphere": The specified datastore: "Datastore2" is not a VSAN datastore. The policy parameters will work only with VSAN Datastore. 
` 

But I think this issue can be closed now.
